### PR TITLE
Quarkus extension: add support for fast-jar

### PR DESCRIPTION
**Pull request by piotr-szybicki**
Hi, qurkus 1.12.0 made building fast jar, default. This PR adds support for that type of packaging (for maven ).

### CODE REVIEW

**Comment by piotr-szybicki** [comment]
I have no idea what the stupid formater wants from me. I run `java -jar google-java-format-1.9-all-deps.jar -r $(find . -name *.java )` on the project.

File: first-party/jib-quarkus-extension-maven/README.md [hunk 1]
```diff
@@ -27,6 +27,9 @@ Note that `<container><mainClass>` should be set to some placeholder value to su
   </dependencies>
 
   <configuration>
+    <!-- set this if you either have quarkus.package.type=fast-jar  -->
+    <!-- or running quarkus version 1.12.0.CR1 or higer (it's default there) -->
+    <fastJarEnabled>true</fastJarEnabled>
```
Comment (chanseokoh) [inline first-party/jib-quarkus-extension-maven/README.md:32]: Since `quarkus.package.type` accepts different types, should we name this `<quarkusPackageType>` and default to `fast-jar`? (Initially, we can make the extension return with an error for other types than `fast-jar` or `jar`.

**Review by chanseokoh — COMMENTED** [review]
Interesting. We are not really following changes in Quarkus, and thanks for letting us know and also taking your time to contribute! This is awesome.

The PR doesn't look like a simple one, and since we don't know well how Quarkus works in this mode, I think we will need quite some time to be able to go over and provide feedback. Please be patient!

For formatting, you can run `./gradlew goJF` (or `./gradlew googleJavaFormat`) to auto-format the code.

And don't forget to follow the CLA bot instructions to sign the CLA. Otherwise, we cannot accept your contribution. :(

File: first-party/jib-quarkus-extension-maven/README.md (hunk 1 above)
Comment (chanseokoh) [reply first-party/jib-quarkus-extension-maven/README.md:32]: Oh, and you should add extension specific options to the extension configuration, not to the main Jib configuration. Here's a [JibLayerFilterExtension example](https://github.com/GoogleContainerTools/jib-extensions/blob/master/first-party/jib-layer-filter-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/layerfilter/Configuration.java) and its [doc](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-maven#examples).

File: first-party/jib-quarkus-extension-maven/README.md [hunk 2]
```diff
@@ -63,3 +66,5 @@ Additionally, overriding the following configurations using Maven and Java syste
 
 - `<container><appRoot>`
 - `<container><jvmFlags>`
+
+Set the `fastJarEnabled` property to true if you like to build your image with fast jar.
```
Comment (chanseokoh) [inline first-party/jib-quarkus-extension-maven/README.md:70]: We don't update the doc until we make a release. This change should be in a separate PR.

File: first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/JibQuarkusExtension.java [hunk 3]
```diff
@@ -177,6 +182,11 @@ private void readJibConfigurations(MavenProject project) {
     if (jibPlugin != null) {
       Xpp3Dom configurationDom = (Xpp3Dom) jibPlugin.getConfiguration();
       if (configurationDom != null) {
+        Xpp3Dom fastJarEnabled = configurationDom.getChild("fastJarEnabled");
+        if (fastJarEnabled != null) {
+          this.fastJarEnabled = Boolean.valueOf(fastJarEnabled.getValue());
+        }
```
Comment (chanseokoh) [inline first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/JibQuarkusExtension.java:188]: As explained above, you don't read extension-specific parameters from the main Jib config.

File: first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/resolvers/DependencyDto.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.maven.extension.quarkus.resolvers;
+
+import java.nio.file.Path;
+import java.util.Optional;
+
+public class DependencyDto {
```
Comment (chanseokoh) [inline first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/resolvers/DependencyDto.java:22]: What does "DTO" stand for?

File: first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/resolvers/FastJarDependencyResolver.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+              + " doesn't exist; did you run the Qaurkus Maven plugin "
+              + "('compile' and 'quarkus:build' Maven goals)?");
+    }
+    return jar;
+  }
+
+  @Override
+  public AbsoluteUnixPath getAppRootJar(AbsoluteUnixPath appRoot) {
+    return appRoot.resolve("quarkus-app/quarkus-run.jar");
+  }
```
Comment (chanseokoh) [inline first-party/jib-quarkus-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/quarkus/resolvers/FastJarDependencyResolver.java:46]: Perhaps
```suggestion
  public Path getRelativeJarPath() {
    return Paths.get("quarkus-app/quarkus-run.jar");
  }
```

**Comment by chanseokoh** [comment]
Still reviewing this. I'll come back to this again when I have time.

**Comment by chanseokoh** [comment]
@piotr-szybicki just cheking in, do you plan to keep working on this? First of all, you need to sign the CLA so that we can accept your contribution.

**Comment by piotr-szybicki** [comment]
> @piotr-szybicki just cheking in, do you plan to keep working on this? First of all, you need to sign the CLA so that we can accept your contribution.

Hi. 1000 Apologies. For some known reason the github notifications ware send to the spam folder. So I didn't notice any of the comments. You mentioned that it might take a long time to review them I will address them over the weekend. As well as figure out how to sign CLA (I'm sure  that I did that. )

**Comment by zanmagerl** [comment]
Is there any update with fast-jar support? @piotr-szybicki do you still intend to work on this or do we try to make it happen in different PR?

**Comment by elefeint** [comment]
Closed as implemented in #129.
